[JSC] add support for `Uint8Array.prototype.toBase64` and `Uint8Array.prototype.toHex`

dcrousso opened the pull request:
#### 2596e190eeb8bcb7062e415c7896c7f875e0d080
<pre>
[JSC] add support for `Uint8Array.prototype.toBase64` and `Uint8Array.prototype.toHex`
<a href="https://bugs.webkit.org/show_bug.cgi?id=274635">https://bugs.webkit.org/show_bug.cgi?id=274635</a>

Reviewed by Yusuke Suzuki.

These methods will make it easier for developers to encode typed character/byte data.

Note that there are other methods in the related proposal, but for ease of reviewing they will be implemented separately.

Spec: &lt;<a href="https://tc39.es/proposal-arraybuffer-base64/spec/">https://tc39.es/proposal-arraybuffer-base64/spec/</a>&gt;
Proposal: &lt;<a href="https://github.com/tc39/proposal-arraybuffer-base64">https://github.com/tc39/proposal-arraybuffer-base64</a>&gt;

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeInlines.h:
(JSC::JSGenericTypedArrayViewPrototype&lt;ViewClass&gt;::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp: Added.
(JSC::uint8ArrayPrototypeToBase64):
(JSC::uint8ArrayPrototypeToHex):

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
Create builtin identifiers for `alphabet` and `omitPadding`.

* Source/JavaScriptCore/runtime/OptionsList.h:
Add a `useUint8ArrayBase64Methods` feature flag for these new methods (and the others that are part of the proposal that haven&apos;t been implemented yet).

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:

* JSTests/stress/uint8array-toBase64.js: Added.
* JSTests/stress/uint8array-toHex.js: Added.

Canonical link: <a href="https://commits.webkit.org/280654@main">https://commits.webkit.org/280654@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e28c23380b0e02e1fbe5d168d351a79bd3d518f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36557 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9704 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60851 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7672 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7862 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46334 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5402 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59259 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/34300 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49410 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27197 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31083 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6726 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6677 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50321 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62530 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56471 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1142 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7089 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53595 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1147 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49449 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12650 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/961 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78231 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12966 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33471 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/34556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33217 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,106 @@
+    if (thisObject->type() != Uint8ArrayType)
+        return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that |this| be a Uint8Array"_s);
+
+    JSUint8Array* uint8Array = jsCast<JSUint8Array*>(thisObject);
```
Let's use `jsDynamicCast<JSUint8Array*>(thisValue)` and check whether it is non null, like
```
JSUint8Array* uint8Array = jsDynamicCast<JSUint8Array*>(callFrame->thisValue());
if (UNLIKELY(!uint8Array))
     return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that |this| be a Uint8Array"_s);
```

dcrousso replied:
nice!  ive always wondered when to use `jsDynamicCast` vs `jsCast`.  is there a general rule i could follow to answer that?

kmiller68 replied:
If you're going to check if a JSValue is some type we typically use `jsDynamicCast`. If you know the JSValue is actually the type already we use `jsCast`. There are other mechanisms but those are mostly for legacy code.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    JSObject* thisObject = jsCast<JSObject*>(thisValue);
+    if (thisObject->type() != Uint8ArrayType)
+        return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toHex requires that |this| be a Uint8Array"_s);
```
Ditto

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,106 @@
+
+        JSObject* optionsObject = jsCast<JSObject*>(optionsValue);
+
+        JSValue alphabetValue = optionsObject->get(globalObject, Identifier::fromString(vm, "alphabet"_s));
```
Let's define `alphabet` in CommonIdentifier

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,106 @@
+    const uint8_t* data = uint8Array->typedVector();
+    size_t offset = uint8Array->byteOffset();
+    size_t length = uint8Array->length();
+    RELEASE_AND_RETURN(scope, JSValue::encode(jsString(vm, base64EncodeToString({ data + offset, length }, mode))));
```
This is not correct. `typedVector` is already offseted vector pointer. `byteOffset` is indicating the offset in the original buffer. Not in the typedVector. Can you add a test for that?

dcrousso replied:
oh!  interesting.  i guess i didnt look at this closely enough as i can clearly see this in `JSArrayBufferView::ConstructionContext` 😅

yes i will add some tests specifically for this scenario

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,106 @@
+    size_t length = uint8Array->length();
+    for (size_t i = offset; i < length; ++i)
+        builder.append(toStringWithRadix(data[i], 16));
+    RELEASE_AND_RETURN(scope, JSValue::encode(jsString(vm, builder.toString())));
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,94 @@
+    const uint8_t* data = uint8Array->typedVector();
+    size_t length = uint8Array->length();
+    for (size_t i = 0; i < length; ++i)
+        builder.append(toStringWithRadix(data[i], 16));
```
This is allocating String per byte, which is too costly.

Jarred-Sumner replied:
Incase it's helpful, in Bun we implemented the Node.js API `Buffer.prototype.toString("hex")`:

https://github.com/oven-sh/bun/blob/aa3aa888d52b8c7546ac365d56a0cd5c990cb7d8/src/string_immutable.zig#L3949-L4038 

It uses SIMD and is [around 10x faster](https://x.com/jarredsumner/status/1693597759310573945/photo/) than Node.js' scalar implementation

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,94 @@
+    validateTypedArray(globalObject, uint8Array);
+
+    StringBuilder builder;
+    builder.reserveCapacity(uint8Array->length());
```
Doesn't one byte become 2 characters? (255 => `ff`)

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -0,0 +1,94 @@
+                mode = Base64EncodeMode::URL;
+            else if (alphabet != "base64"_s)
+                return throwVMTypeError(globalObject, scope, "Uint8Array.prototype.toBase64 requires that alphabet be \"base64\" or \"base64url\""_s);
+        }
```
These things have side-effect since it is invoking `toString`, `[[Get]]`, so they can detach the typed array.
Doesn't the spec checks the condition after that?

dcrousso commented on JSTests/stress/uint8array-toBase64.js:
```diff
@@ -0,0 +1,126 @@
+}
+
+shouldBe((new Uint8Array([])).toBase64({alphabet: "base64url"}), "");
+shouldBe((new Uint8Array([0])).toBase64({alphabet: "base64url"}), "AA");
```
i just realized these are all missing padding `=`, which im fixing as a prerequisite in <https://github.com/WebKit/WebKit/pull/29530> (which should also help if/when `omitPadding` is approved in TC39 plenary later this month)

Constellation reviewed (approved):
r=me

dcrousso committed:
[JSC] add support for `Uint8Array.prototype.toBase64` and `Uint8Array.prototype.toHex`
https://bugs.webkit.org/show_bug.cgi?id=274635

Reviewed by Yusuke Suzuki.

These methods will make it easier for developers to encode typed character/byte data.

Note that there are other methods in the related proposal, but for ease of reviewing they will be implemented separately.

Spec: <https://tc39.es/proposal-arraybuffer-base64/spec/>
Proposal: <https://github.com/tc39/proposal-arraybuffer-base64>

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeInlines.h:
(JSC::JSGenericTypedArrayViewPrototype<ViewClass>::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp: Added.
(JSC::uint8ArrayPrototypeToBase64):
(JSC::uint8ArrayPrototypeToHex):

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
Create builtin identifiers for `alphabet` and `omitPadding`.

* Source/JavaScriptCore/runtime/OptionsList.h:
Add a `useUint8ArrayBase64Methods` feature flag for these new methods (and the others that are part of the proposal that haven't been implemented yet).

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:

* JSTests/stress/uint8array-toBase64.js: Added.
* JSTests/stress/uint8array-toHex.js: Added.

Canonical link: https://commits.webkit.org/280654@main
